### DISCUSSION TITLE
Styled JSX - Vercel

### DISCUSSION THREAD

**Discussion by vczb** [frontendbr/forum#2007] [Discussão]
O que vocês acham de [styled jsx](https://github.com/vercel/styled-jsx) ?, peguei um exemplo de um [blog](https://github.com/wutali/nextjs-netlify-blog-template)

Dizem que tem vantagens de performance etc etc... Mmas olhando parece bem pouco prático na verdade... Quero aprender mais, alguém conhece ?

```
import Head from "next/head";
import Navigation from "./Navigation";

type Props = {
  children: React.ReactNode;
};
export default function Layout({ children }: Props) {
  return (
    <div className="root">
      <Head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link rel="manifest" href="/site.webmanifest" />
        <link rel="apple-touch-icon" href="/icon.png" />
        <meta name="theme-color" content="#fff" />
      </Head>
      <nav>
        <Navigation />
      </nav>
      <main>{children}</main>
      <style jsx>
        {`
          .root {
            display: block;
            padding: 4rem 0;
            box-sizing: border-box;
            height: 100%;
          }
          main {
            display: flex;
            min-height: 100%;
          }
          @media (min-width: 769px) {
            .root {
              display: flex;
              flex: 1 0 auto;
            }
            main {
              flex: 1 0 auto;
            }
          }
        `}
      </style>
    </div>
  );
}
```

**Comment by giovannism20** [comment]
> O que vocês acham de [styled jsx](https://github.com/vercel/styled-jsx) ?, peguei um exemplo de um [blog](https://github.com/wutali/nextjs-netlify-blog-template)
> 
> Dizem que tem vantagens de performance etc etc... Mmas olhando parece bem pouco prático na verdade... Quero aprender mais, alguém conhece ?
> 
> ```
> import Head from "next/head";
> import Navigation from "./Navigation";
> 
> type Props = {
>   children: React.ReactNode;
> };
> export default function Layout({ children }: Props) {
>   return (
>     <div className="root">
>       <Head>
>         <meta charSet="utf-8" />
>         <meta name="viewport" content="width=device-width, initial-scale=1" />
>         <link rel="manifest" href="/site.webmanifest" />
>         <link rel="apple-touch-icon" href="/icon.png" />
>         <meta name="theme-color" content="#fff" />
>       </Head>
>       <nav>
>         <Navigation />
>       </nav>
>       <main>{children}</main>
>       <style jsx>
>         {`
>           .root {
>             display: block;
>             padding: 4rem 0;
>             box-sizing: border-box;
>             height: 100%;
>           }
>           main {
>             display: flex;
>             min-height: 100%;
>           }
>           @media (min-width: 769px) {
>             .root {
>               display: flex;
>               flex: 1 0 auto;
>             }
>             main {
>               flex: 1 0 auto;
>             }
>           }
>         `}
>       </style>
>     </div>
>   );
> }
> ```

Acho que esse exemplo é um tanto quanto ruim mas para te responder por agora vou te deixar esse link: [8 reasons to use styled-components](https://blog.logrocket.com/8-reasons-to-use-styled-components-cf3788f0bb4d/) - Ele vai te trazer diversos exemplos práticos.

**Reply by vczb** [reply]
Styled-Components eu conheço e gosto bastante! estou me referindo ao styled jsx da vercel

https://github.com/vercel/styled-jsx

aqui tem mais um exemplo dele

```
export default () => (
  <div className="root">
    <style jsx>{`
      .root {
        color: green;
      }
    `}</style>
  </div>
)
```

É tipo um css inline... vc já usou ?

**Reply by vczb** [reply]
Olha como eles declaram um botão:

```
const Button = (props) => (
  <button>
     { props.children }
     <style jsx>{`
        button {
          color: #999;
          display: inline-block;
          font-size: 2em;
        }
     `}</style>
     <style jsx>{`
        button {
          padding: ${ 'large' in props ? '50' : '20' }px;
          background: ${props.theme.background};
        }
     `}</style>
  </button>
)
```

**Comment by ninetails** [comment]
Não tenho certeza mas talvez tenha uma otimização caso você utilizar [Next.js](https://nextjs.org/), pois são os mesmos mantenedores e vem ativo por padrão neste _metaframework_ (mas não achei nada falando sobre, explicitamente).

Mas pessoalmente ainda prefiro utilizar um [styled-components](https://styled-components.com/) ou [emotion](https://emotion.sh/)

**Comment by didiraja** [comment]
Styled Components é ótimo, mas pra mim só fez sentido dentro do ambiente React (nunca mexi com Angular). Com Svelte e Vue, prefiro csszinho mesmo, turbinado com pre-processor.

**Comment by vczb** [comment]
Quanto a esse [jsx](https://github.com/vercel/styled-jsx) que a pessoa enfia o css ali no meio do html fico me perguntando porque usar ele ao invés de um sass com classnames ?! a única resposta que me vem em mente é perfomance. Imagino que o  [jsx](https://github.com/vercel/styled-jsx) combinado com algum style guide tipo tailwind, bootstrap , material  ou outro, faça mais sentido, mas ele sozinho... ainda não me convenci

**Reply by ninetails** [reply]
Na verdade a melhor forma de utilizá-lo em todo potencial seria sem nenhum framework css externo. A maioria das implementações `styled-component` fazem com que o CSS gerado fique _escopado_ dentro de onde está sendo utilizado, evitando assim que outros elementos da página herdem por cascata suas propriedades.

E a maioria deles também possibilita a interpolação de dados, ou seja, é possível no React passar uma _prop_ e utilizá-la dentro do próprio CSS. Como os estilos são gerados de forma dinâmica na maioria das bibliotecas, o CSS não fica preso a um valor predeterminado pois é gerado em _runtime_.

Sim, deve haver uma perda de performance pois há um processamento na máquina cliente, mas geralmente o ganho desse dinamismo tende a ser maior para a maioria dos casos.
